platform/core/lib/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+pub mod core {
+    use failure::CoreFailure;
+    use feature_categories;
+    use io::{file_system::DefaultFileSystem, key_value_store::DefaultKeyValueStore};
+    use preferences::Preferences;
+    use user_session::{TimeoutOption, UserSession, UserSessionTimeout};
+
+    use once_cell::sync::OnceCell;
+    use serde::{Deserialize, Serialize};
+    use std::sync::{Arc, Mutex};
+    use std::{sync::MutexGuard, time::Instant};
+
+    #[cfg(target_os = "android")]
+    use {
+        android_logger::Config,
+        log::{trace, Level},
+    };
+
+    // Core is held as a singleton.
+    static CORE: OnceCell<Mutex<Core>> = OnceCell::new();
+
+    #[derive(Debug, Clone, Serialize)]
+    pub enum PlatformRequest {
+        Example,
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    pub enum PlatformResponse {
+        Example(String),
+    }
+
+    pub trait PlatformHookRequest: Sync + Send {
+        fn perform_request(
+            &self,
+            request: PlatformRequest,
+        ) -> Result<PlatformResponse, CoreFailure>;
+    }
+
+    const PREFERENCES_DB: &str = "preferences_db";
+
+    // The Core would act as a composition root, containing any global configuration
+    // to be shared between components, as well managing components lifetime.
+    struct Core<'a> {
+        language_code: String,
+        #[allow(dead_code)]
+        preferences: Arc<Preferences>,
+        user_session: Mutex<UserSession<'a>>,
+        #[allow(dead_code)]
+        platform_hook: Box<dyn PlatformHookRequest>,
+    }
+
+    fn get_instance() -> Result<MutexGuard<'static, Core<'static>>, CoreFailure> {
+        match CORE.get() {
+            Some(core) => core
+                .lock()
+                .map_err(|_| CoreFailure::core_already_bootstrapped()),
+            None => Err(CoreFailure::core_not_bootstrapped()),
+        }
+    }
+
+    pub fn bootstrap(
+        language_code: String,
+        fs_root: String,
+        platform_hook: Box<dyn PlatformHookRequest>,
+    ) -> Result<(), CoreFailure> {
+        if CORE.get().is_some() {
+            return Err(CoreFailure::core_already_bootstrapped());
+        }
+        let preferences = Arc::new(Preferences {
+            store: Box::new(DefaultKeyValueStore::new(fs_root + "/" + PREFERENCES_DB)),
+        });
+
+        let builder = Box::new(Instant::now);
+        let user_session = Mutex::from(UserSession::new(builder, preferences.clone()));
+        let core = Core {
+            language_code,
+            preferences,
+            user_session,
+            platform_hook,
+        };
+
+        let _ = CORE.set(Mutex::from(core));
+
+        #[cfg(target_os = "android")]
+        {
+            android_logger::init_once(Config::default().with_min_level(Level::Trace));
+            trace!("Rust:core => Bootstrapped!");
+        }
+
+        // For testing purposes
```

Comment:
... should be uncommented?

Reply:
It is meant for local testing purposes, probably needs to be deleted.